### ISSUE TITLE
NumbaTypeError: Cannot modify readonly array of type: readonly array(int64, 2d, C)

### ISSUE THREAD

**Issue by tramchamploo** [numba/numba#8370]
```
import numpy as np
from numba import njit

@njit
def unfold(x, y):
    r, c = x.shape
    a = np.broadcast_to(x, (y, r, c))
#     a = a.copy()
    b = np.swapaxes(a, 0, 1)
#     b = b.copy()
    cc = b.flatten()
#     c = c.copy()
    d = np.reshape(cc, (-1, c))
    # assign indicator to the right end of the period
    # in case of look forward bias
#     d = d.copy()
    d[y - 1 :, :] = d[: 1 - y]
    return d


unfold(np.array([[1, 1, 1], [2, 2, 2], [3,3,3]]), 2)
```
This will reproduce the error above. Under plain numpy this works fine.
By uncommenting any one of the copy() calls, the issue is resolved. So I have no idea which numpy call returns this readonly array.
Can it be implemented without an explicit copy call?

**Comment by gmarkall** [comment]
Thanks for the issue report and nice reproducer! I can reproduce the issue.

The first `readonly array` is `a`, so copying any of the subsequent values breaks the "chain" of readonly arrays.

**Comment by gmarkall** [comment]
Looks like `array.flatten` erroneously types its result as `readonly` if its input is `readonly`, which is unnecessary as it returns a copy. With:

```diff
diff --git a/numba/core/typing/arraydecl.py b/numba/core/typing/arraydecl.py
index 629263aad..edb4ca103 100644
--- a/numba/core/typing/arraydecl.py
+++ b/numba/core/typing/arraydecl.py
@@ -481,7 +481,7 @@ class ArrayAttribute(AttributeTemplate):
         # Only support no argument version (default order='C')
         assert not kws
         assert not args
-        return signature(ary.copy(ndim=1, layout='C'))
+        return signature(ary.copy(ndim=1, layout='C', readonly=False))
 
     @bound_function("array.take")
     def resolve_take(self, ary, args, kws):
```

the following:

```python
import numpy as np
from numba import njit


def unfold(x, y):
    r, c = x.shape
    a = np.broadcast_to(x, (y, r, c))
#     a = a.copy()
    b = np.swapaxes(a, 0, 1)
#     b = b.copy()
    cc = b.flatten()
#     c = c.copy()
    d = np.reshape(cc, (-1, c))
    # assign indicator to the right end of the period
    # in case of look forward bias
#     d = d.copy()
    d[y - 1:, :] = d[: 1 - y]
    return d


x = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])

jitted = njit(unfold)

print(unfold(x, 2))
print(jitted(x, 2))
```

gives

```
$ python repro.py 
[[1 1 1]
 [1 1 1]
 [1 1 1]
 [2 2 2]
 [2 2 2]
 [3 3 3]]
[[1 1 1]
 [1 1 1]
 [1 1 1]
 [2 2 2]
 [2 2 2]
 [3 3 3]]
```

as one would expect.

To fix this, the above change can be seen as a starter patch that needs test cases (probably including the reproducer from this report) adding. This may be a good first issue for a new contributor to take on, if a maintainer doesn't get round to this first, so I'll label it as such,

**Comment by tramchamploo** [comment]
So you mean broadcast_to producing a readonly array is the expected behaviour, but under plain numpy even replacing `flatten` to `ravel` works which in the numpy document says `ravel` does not necessarily copy the array. So there could be some other unconsistency in behaviour of these calls between numba and numpy?

**Comment by gmarkall** [comment]
> So you mean broadcast_to producing a readonly array is the expected behaviour

My current understanding is that this is a bug, and it shouldn't be readonly.

>  but under plain numpy even replacing `flatten` to `ravel` works which in the numpy document says `ravel` does not necessarily copy the array. So there could be some other unconsistency in behaviour of these calls between numba and numpy?

I'm not quite sure what you're suggesting here - if it seems like there's another NumPy-Numba inconsistency in `ravel`, could you please open a new issue that reproduces the inconsistency?

**Comment by tramchamploo** [comment]
Well, I thought you were suggesting `flatten` is the cause. My fault, your last post make this clear. I didn't dig into numpy's source code and just making a rough judge from its document. Will see what I can do on this when I have time.

**Comment by gmarkall** [comment]
Sorry, I may be being confusing - I do think `flatten` is the cause, but the error in `flatten` is in the Numba implementation of it - I don't suspect anything wrong with NumPy itself.

**Comment by gmarkall** [comment]
(For clarity, I also don't think there's anything wrong with your code - that looks like it should work)

**Comment by tramchamploo** [comment]
Since you think `flatten` is the cause, then `ravel` is not expected to work. But in numpy `flatten` and `ravel` all work fine. And in your PR, fixing `flatten` alone seems to have an impact on `ravel` as well. I'm not familiar with the compiler side of numba and a little bit confused now.

**Comment by bszollosinagy** [comment]
Yes, in NumPy both work fine. While making the fix for this issue, I also looked at `ravel` because that one also flattens the array, and also makes a copy if needed. So I was curious if Numba behaved exactly like NumPy and I found that `ravel` also had a little problem. Ideally, that fix could go into a separate issue, and a separate Pull Request, so now it's simply piggy-back riding here, due to its similarity of cause.
